### ISSUE TITLE
Problem on function "extract_patches"

### ISSUE THREAD

**Issue by facetohard** [WonwoongCho/Generative-Inpainting-pytorch#2]
I found that the reimplementation of "extract_patches" may be wrong. 
I tested your context_attention module according to function "[test_contextual_attention](https://github.com/JiahuiYu/generative_inpainting/blob/master/inpaint_ops.py)". But i got the different result.
I changed it to 
`x = self.padding(x)`
`x = x.permute(0, 2, 3, 1)`
 `all_patches = x.unfold(1, kernel, stride).unfold(2, kernel, stride) `
and I got the same results with the tf code.
May be this difference results the blur results you have mentioned in Readme

**Comment by robbiebarrat** [comment]
@facetohard which file did you edit? `model_module.py`?

If you could please let me know which file / line numbers to edit - or perhaps just put the whole new file in a gist, it would really help me out. I'm getting blurry results too and badly want this fix!!

Thank you!!

**Comment by facetohard** [comment]
Line 263 in [model_module.py](https://github.com/WonwoongCho/Generative-Inpainting-pytorch/blob/master/model_module.py)

**Comment by hedlmote** [comment]
> I found that the reimplementation of "extract_patches" may be wrong.
> I tested your context_attention module according to function "[test_contextual_attention](https://github.com/JiahuiYu/generative_inpainting/blob/master/inpaint_ops.py)". But i got the different result.
> I changed it to
> `x = self.padding(x)`
> `x = x.permute(0, 2, 3, 1)`
> `all_patches = x.unfold(1, kernel, stride).unfold(2, kernel, stride) `
> and I got the same results with the tf code.
> May be this difference results the blur results you have mentioned in Readme

When I check the implementation, I also find that the extract_patches is wrong. The authors directly unfold background tensors and view the extract patch as shape of weight size needed for convolution. Though running is ok, but such operation by no means is to apply cross-correlation between each background patch and the the foreground tensor.

**Comment by Zealoe** [comment]
can you get better results now with this fix? Thanks!

**Comment by hedlmote** [comment]
The discriminator and the l2_norm also acts differently with the original implementation hedl 邮箱：hedl@mail.ustc.edu.cn 签名由 网易邮箱大师 定制 在2019年04月30日 19:38，Zealoe 写道： can you get better results now with this fix? Thanks! — You are receiving this because you commented. Reply to this email directly, view it on GitHub, or mute the thread.

**Comment by Zealoe** [comment]
Do you mean that contextual attention module is right with this fix now? 
Thanks a lot! @hedlmote

**Comment by hedlmote** [comment]
yes, i think so. hedl 邮箱：hedl@mail.ustc.edu.cn 签名由 网易邮箱大师 定制 在2019年05月01日 20:41，Zealoe 写道： Do you mean that contextual attention module is right with this fix now? Thanks a lot! @hedlmote — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub, or mute the thread.

**Comment by baogiadoan** [comment]
> I found that the reimplementation of "extract_patches" may be wrong.
> I tested your context_attention module according to function "[test_contextual_attention](https://github.com/JiahuiYu/generative_inpainting/blob/master/inpaint_ops.py)". But i got the different result.
> I changed it to
> `x = self.padding(x)`
> `x = x.permute(0, 2, 3, 1)`
> `all_patches = x.unfold(1, kernel, stride).unfold(2, kernel, stride) `
> and I got the same results with the tf code.
> May be this difference results the blur results you have mentioned in Readme

If you think it's a patch, I think you should submit a pull request for the author to update it, otherwise people will use a bug version, thanks. I'll try yours to see the result.

**Comment by ShnitzelKiller** [comment]
I've been getting an all gray image when running test_contextual_attention from the TF framework, even after applying the patch described here. For running test_contextual_attention, you are using just the output of the conv_transpose2d, without the last self.out(y), right?

Here are some example result comparisons with a test image (where F and B are both the same image, and mask is None):
Input:
![testimg_large](https://user-images.githubusercontent.com/6132502/58361020-b3240600-7e40-11e9-9db0-865200bc7163.png)

Result with tensorflow implementation (baseline):
![result_tf3](https://user-images.githubusercontent.com/6132502/58361042-ce8f1100-7e40-11e9-9fc3-f0b3418f678d.png)

Result with this framework, with the patch to fix extract_patches:
![result](https://user-images.githubusercontent.com/6132502/58361056-e5cdfe80-7e40-11e9-8d94-5d4353054040.png)

EDIT:
I finally narrowed down the bug, and it seems that the down_sample() function in utils.py is completely broken. The downsampled image loses all of the information and doesn't match the original at all. I switched it to the built in nn.functional.interpolate() and then it worked.

EDIT 2:
I found another bug that was preventing things from working properly. It's the normalization of the convolution patches:
```python
wi_normed = wi / torch.max(l2_norm(wi), escape_NaN)
```
I replaced it with the equivalent code in the tensorflow implementation (normalizing per kernel patch):
```python
wi_normed = wi / torch.max(torch.sqrt((wi*wi).sum([1,2,3],keepdim=True)), escape_NaN)
```

And NOW it actually works. Without this change, I would still get a flat color image sometimes.
